### PR TITLE
Fast path in SET if the expiration time is expired

### PR DESCRIPTION
**Pull request by enjoy-binbin**
If the expiration time passed in SET is expired, for example, it
has expired due to the machine time (DTS) or the expiration time
passed in (wrong arg). In this case, we don't need to set the key
and wait for the active expire scan before deleting the key.

Compared with previous changes:
1. If the key does not exist, previously we would set the key and wait
for the active expire to delete it, so it is a set + del from the perspective
of propaganda. Now we will no set the key and return, so it a NOP.

2. If the key exists, previously we woule set the key and wait
for the active expire to delete it, so it is a set + del From the perspective
of propaganda. Now we will delete it and return, so it is a del.

Adding a new deleteExpiredKeyFromOverwriteAndPropagate function
to reduce the duplicate code.

A simple benchmark test:
```
./src/valkey-benchmark -P 32 -c 100 -n 10000000 -r 1000000000 set __rand_int__ value exat 100

old: 252474.23, new: 829256.12
```

### CODE REVIEW

File: src/t_string.c [hunk 1]
```diff
@@ -109,6 +109,28 @@ void setGenericCommand(client *c,
         return;
     }
 
+    /* If the milliseconds have expired, then we don't need to set it into the
```
Comment (hpatro) [inline src/t_string.c]: This sentence didn't make sense immediately. Could we add back tick to the variable name like `milliseconds`

File: src/t_string.c [hunk 2]
```diff
@@ -109,6 +109,28 @@ void setGenericCommand(client *c,
...
+            int deleted = dbGenericDelete(c->db, key, server.lazyfree_lazy_expire, DB_FLAG_KEY_EXPIRED);
+            serverAssertWithInfo(c, key, deleted);
+            server.dirty++;
+
+            /* Replicate/AOF this as an explicit DEL or UNLINK. */
+            robj *aux = server.lazyfree_lazy_expire ? shared.unlink : shared.del;
+            rewriteClientCommandVector(c, 2, aux, key);
+            signalModifiedKey(c, c->db, key);
+            notifyKeyspaceEvent(NOTIFY_GENERIC, "del", key, c->db->id);
+        }
```
Comment (hpatro) [inline src/t_string.c]: This code flow is duplicated from expire.c could we abstract this out to server.c maybe and invoke the same from expire.c / t_string.c ?

Comment (enjoy-binbin) [reply src/t_string.c]: i do add a new function to reduce the duplicated in the very first version: https://github.com/valkey-io/valkey/commit/2b0295623. But i struggled with the name and don't like the diff very much, so came up with this.

Do you think the diff is better? i am seeking more votes.

Comment (madolson) [reply src/t_string.c]: I like the abstraction, and would would name it something like `deleteExpiredKeyFromOverwriteAndPropagate`.

**Review by madolson — DISMISSED** [review]

**Comment by enjoy-binbin** [comment]
ok, the test failed because the old restore use `lazyfree_lazy_server_del` to delete the expired key (not the lazy expire)

i am going to leave store alone.

File: tests/unit/type/string.tcl [hunk 3]
```diff
@@ -607,6 +607,40 @@ if {[string match {*jemalloc*} [s mem_allocator]]} {
         r set foo bar pxat [expr [clock milliseconds] + 10000]
         assert_range [r ttl foo] 5 10
     }
+
+    test "SET EXAT / PXAT Expiration time is expired" {
```
Comment (hpatro) [inline tests/unit/type/string.tcl:611]: Is there test coverage for `unlink` client rewrite scenario already?

**Review by hpatro — APPROVED** [review]

File: tests/unit/type/string.tcl (hunk 3 above)
Comment (enjoy-binbin) [reply tests/unit/type/string.tcl:611]: i guess there is no coverage for unlink, but the code is easy so think this is enough.

File: src/db.c [hunk 4]
```diff
@@ -1712,6 +1712,21 @@ void deleteExpiredKeyAndPropagate(serverDb *db, robj *keyobj) {
     server.stat_expiredkeys++;
 }
 
+/* Delete the specified expired key from overwriting and propagate the DEL or UNLINK. */
+void deleteExpiredKeyFromOverwriteAndPropagate(client *c, robj *keyobj, int do_delete) {
```
Comment (zuiderkwast) [inline src/db.c]: The difference between this function and deleteExpiredKeyAndPropagate is not very clear to me. When is each of these used? Maybe the name can be better or we can rename both functions if we can find a more clear purpose of each one with other names.

"From Overwrite" means that an explicit command that deletes the key?

It's a bit confusing to have a `do_delete` that can be false in a function that is called delete-something. Can we avoid it? What I can see, this function is always called with `do_delete = 1`.

Comment (enjoy-binbin) [reply src/db.c]: ohh, the do_delete use was removed in here: https://github.com/valkey-io/valkey/pull/865/commits/e9e089bcf4134b4add98616a01c726d3de943782

i forgot to remove it.

The `deleteExpiredKeyFromOverwriteAndPropagate` name was suggested by Madelyn

File: src/cluster.c [hunk 5]
```diff
@@ -262,6 +262,8 @@ void restoreCommand(client *c) {
     if (ttl && !absttl) ttl += commandTimeSnapshot();
     if (ttl && checkAlreadyExpired(ttl)) {
         if (deleted) {
+            /* Here we don't use deleteExpiredKeyFromOverwriteAndPropagate because
+             * strictly speaking, the `deleted` is triggered by `replace`. */
```
Comment (madolson) [inline src/cluster.c]:
```suggestion
             * strictly speaking, the `delete` is triggered by the `replace`. */
```

**Review by madolson — APPROVED** [review]
Saw there was some back and forth about the specifics, but the latest version looks good to me outside of a minor wording suggestion.

**Review by zuiderkwast — APPROVED** [review]

**Review by lilpoozie2005 — APPROVED** [review]

**Pull request merged**